apps/middleman/src/actions/Providers.ts

```diff
@@ -19,26 +22,122 @@ const updateProvidersSchema = z.object({
   }),
 });
 
-export async function UpdateProvidersFromSource(): Promise<Provider[]> {
-  const applicationSettings = await getApplicationSettings();
+export async function UpdateProvidersFromSource() {
+  const [userIdentity, appSettings] = await Promise.all([
+    getCurrentUserIdentity(),
+    getApplicationSettings(),
+  ]);
 
-  const url = process.env.PROVIDERS_CDN_URL!.replace('{chainId}', applicationSettings.chainId);
+  const providersCdnUrl = process.env.PROVIDERS_CDN_URL!.replace(
+    "{chainId}",
+    appSettings.chainId,
+  );
 
-  if (!url) {
-    throw new Error("PROVIDERS_CDN_URL is not defined");
+  if (!providersCdnUrl) {
+    throw new Error("PROVIDERS_CDN_URL environment variable is not defined");
   }
 
+  console.log(`[Providers] Starting update from ${providersCdnUrl}`);
+
   try {
-    const response = await fetch(url);
+    const response = await fetch(providersCdnUrl);
+
     if (!response.ok) {
-      throw new Error(
-        `Failed to fetch providers from CDN: ${response.statusText}`
-      );
+      throw new Error(`Failed to fetch providers: ${response.statusText}`);
+    }
+
+    type CdnProvider = {
+      name: string;
+      identity: string;
+      identityHistory: string[];
+      url: string;
+    };
+
+    const providersFromCdn = (await response.json()) as CdnProvider[];
+    console.log(`[Providers] Fetched ${providersFromCdn.length} providers from CDN`);
+
+    const currentProviders = await list(true);
+
+    const currentProvidersMap = new Map(
+      currentProviders.map((p) => [p.identity, p]),
+    );
+
+    const allCdnIdentities = new Set<string>();
+
+    let inserted = 0;
+    let updated = 0;
+    let disabled = 0;
+
+    for (const p of providersFromCdn) {
+      allCdnIdentities.add(p.identity);
+      p.identityHistory.forEach((h) => allCdnIdentities.add(h));
+    }
+
+    for (const cdnProvider of providersFromCdn) {
+      const possibleIds = [cdnProvider.identity, ...cdnProvider.identityHistory];
+
+      const matchingCurrent =
+        possibleIds.map((id) => currentProvidersMap.get(id)).find(Boolean) ??
```

Comment:
Is not better to simply do:

```typescript
const matchingId = possibleIds.find((id) => !!currentProvidersMap.get(id));
const matchingCurrent = currentProvidersMap.get(matchingId)
```

Or a for loop that sets the variable `matchingCurrent` if `currentProvidersMap.get` call is expensive.